File: problems/cpp/3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int ans = 100000;
        int n = nums.size();
        sort(nums.begin(), nums.end());
        for(int i = 0; i < n; i++){
            for(int j = i + 1; j < n; j++){
                int pos = lower_bound(nums.begin(), nums.end(), target - nums[i] - nums[j]) - nums.begin();
                for(int k = -3; k < 3; k++){
                    int p = pos + k;
                    if(0 <= p && p < n && p != i && p != j && abs(ans - target) > abs(nums[i] + nums[j] + nums[p] - target)){
                        ans = nums[i] + nums[j] + nums[p];
                    }
                }
            }
        }
        return ans;
    }
};

File: problems/python/3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        ans = sum(nums[:3])
        nums.sort()
        n = len(nums)
        for i in range(n):
            for j in range(i + 1, n):
                need = target - nums[i] - nums[j]
                pos = bisect.bisect_left(nums, need)
                p = pos
                while p == i or p == j:
                    p -= 1
                if 0 <= p < n and abs(ans - target) > abs(nums[i] + nums[j] + nums[p] - target):
                    ans = nums[i] + nums[j] + nums[p]
                
                p = pos - 1
                while p == i or p == j:
                    p -= 1
                if 0 <= p < n and abs(ans - target) > abs(nums[i] + nums[j] + nums[p] - target):
                    ans = nums[i] + nums[j] + nums[p]
                
                p = pos + 1
                while p == i or p == j:
                    p += 1
                if 0 <= p < n and abs(ans - target) > abs(nums[i] + nums[j] + nums[p] - target):
                    ans = nums[i] + nums[j] + nums[p]
        return ans